File: LZW/LZW_encode_binary.py
# -*- coding: utf-8 -*-
import math
with open('LZW_toEncode.txt', 'r') as f:
	s = f.read()
d = [chr(i) for i in range(256)]

stack = ''
flow = []

for c in s:
	stack += c
	if stack not in d:
		flow.append(bin(d.index(stack[:-1]))[2:].rjust(int(math.log(len(d), 2))+1, '0'))
		d.append(stack)
		stack = c

if stack:
	flow.append(bin(d.index(stack))[2:].rjust(int(math.log(len(d), 2))+1, '0'))

with open('LZW_Encoded.txt', 'w') as F:
	F.write(''.join(map(str, flow)))

File: LZW/LZW_encode_binary.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <string>
#include <cmath>

#define pb push_back

using namespace std;

string convert2bits(long long n, long long size) {
	string s;
	for (int i=size-1; i>=0; --i) {
		long long p = pow(2, i);
		s += (n / p)?"1":"0";
		n %= p;
	}
	return s;
}

unsigned char convert2byte(string s) {
	int size = s.length();
	unsigned char b = 0;
	for (int i=size; i>0; --i) {
		b += (int)pow(2, i-1);
	}
	return b;
}

int main()
{
	ifstream fi("LZW_toEncode.txt");

	if (fi.is_open()) {
		string flow;
		char c;
		string stack;
		string newstack;
		map<string, int> d;
		for (int i=0; i<256; ++i) {
			d[string(1, i)] = i;
		}
		int last = 256;

		while (fi.get(c)) {
			newstack = stack + c;
			if (d.find(newstack) == d.end()) {
				flow += convert2bits(d[stack], (long long)log2(last)+1);
				d[newstack] = last++;
				newstack = c;
			}
			stack = newstack;
		}
		flow += convert2bits(d[stack], (long long)log2(last)+1);

		ofstream fo("LZW_Encoded.txt", ios::binary);

		if (fo.is_open()) {
			fo << flow;
			fo.close();
		} else {
			cout << "Unable to open output file :C" << endl;
		}
		fi.close();
	} else {
		cout << "Unable to open input file :c" << endl;
	}
	return 0;
}
